### ISSUE TITLE
Make tests more reliable

### ISSUE THREAD

**Issue by paulmelnikow** [badges/shields#911]
I've noticed a few problems with the tests. 

I have this test failing silently on my machine. 
```js
['should produce PNG badges', function(done, assert) {
  console.log(1);
  child = cproc.spawn('node',
    ['test/cli-test.js', 'cactus', 'grown', '.png']);
  child.stdout.on('data', function(chunk) {
    console.log(2);
    // Check the PNG magic number.
    assert.equal(chunk[0], 0x89);
    assert.equal(chunk[1], 0x50);
    assert.equal(chunk[2], 0x4e);
    assert.equal(chunk[3], 0x47);
    assert.equal(chunk[4], 0x0d);
    assert.equal(chunk[5], 0x0a);
    assert.equal(chunk[6], 0x1a);
    assert.equal(chunk[7], 0x0a);
    done();
  });
```

`1` prints but `2` doesn't, then the test suite finishes without an error or any other output.

Whether or not it prints a stack trace seems to depend on timing. Sometimes I get the stack trace, sometimes I don't.

Some of the CI builds appear to be generating uncaught exceptions, yet the CI build is passing.

Here is one example:

- https://travis-ci.org/badges/shields/builds/215764775

```sh
> gh-badges@1.3.0 test:js /home/travis/build/badges/shields
> node test/test.js
No secret data (private/secret.json, see server.js): { Error: Cannot find module './private/secret.json'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/travis/build/badges/shields/server.js:36:19)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3) code: 'MODULE_NOT_FOUND' }
No secret data (private/secret.json, see server.js): { Error: Cannot find module '../private/secret.json'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/travis/build/badges/shields/lib/suggest.js:7:19)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3) code: 'MODULE_NOT_FOUND' }
```

I think it started before that commit though. I'm trying to track it down but Travis is being slow.

**Added:**

Here's a build of an earlier commit with the same issue: https://travis-ci.org/badges/shields/builds/215308677

**Comment by paulmelnikow** [comment]
@espadrine We could debug the test harness, though it might be more time-effective to use one that's already proven. I've had good experiences with [Mocha](https://mochajs.org/). It has good support for callbacks, promises, and synchronous tests.

I worked up a quick POC in a branch and it catches the silent failures on my machine. (They register as timeouts.)

**Comment by paulmelnikow** [comment]
I opened a pull request: #915

**Comment by paulmelnikow** [comment]
Ah, it seems those errors in CI are being emitted from a catch block.

Nevertheless, it seems there's currently no guarantee that `done()` is invoked before the tests return.

**Comment by espadrine** [comment]
What is the error you have?

The `No secret data` is perfectly normal if you don't have a secret file, although I know this error is very disconcerting (I added it for sysadmin purposes). It should be improved (somewhat related to #894).

**Comment by paulmelnikow** [comment]
Yes, agreed about "No secret data".

I wasn't getting an error. The test was failing silently. The tests passed, even though the CLI generated no output and the `.on('data')` block was never invoked.

**Comment by paulmelnikow** [comment]
Resolved in #915.
